File: Pramp/islandCount.js
/**
 * Created by jsonmartin on 6/2/16.
 */
"use strict";

function parseStringLeetCode(matrix) {
  let parsedStringInput = [];
    console.log("Matrix Before trans:", matrix);

    matrix.forEach((row) => {
      if(!Array.isArray(row)) {
        let newRow = row.split("").map((num) => parseInt(num));
        parsedStringInput.push(newRow);
        console.log("parsedStringInput:", parsedStringInput)
      } else {
        console.log("Row:", row);
        let newRow = row.map((num) => parseInt(num));
        parsedStringInput.push(newRow);
      }
    });

    console.log("Matrix after trans:", parsedStringInput);
    return parsedStringInput;
}

function islandCount(matrix) {
  matrix = parseStringLeetCode(matrix); // This helper method transforms input from LeetCode

  let islandCount = 0;

  let checkSingleArrayRow = (col) => { // Added for LeetCode single row test case
    if(col < 0 || col >= matrix.length) { return; } //Base Case
    if(matrix[col] === 1) {
      matrix[col] = 2;
      checkSingleArrayRow(col - 1); // Check Left
      checkSingleArrayRow(col + 1); // Check Right
    }
  };

  let checkEdges = (row, col) => {
    // Check bounds
    if(row < 0 || row >= matrix.length) { return; } //Base Case
    if(col < 0 || col >= matrix[0].length) { return; } //Base Case

    // Traverse Each Cardinal direction in matrix, checking for 1s
    if(matrix[row][col] === 1) {
      matrix[row][col] = 2;
      checkEdges(row, col - 1); // Check Left
      checkEdges(row, col + 1); // Check Right
      checkEdges(row + 1, col); // Check Up
      checkEdges(row - 1, col); // Check Down
    }
  };

  if(!Array.isArray(matrix[0])) { // If only single row matrix (array), LeetCode addition
    for(let col = 0; col < matrix.length; col++) {
      //console.log(matrix);
      if(matrix[col] === 1) {
        islandCount++;
        checkSingleArrayRow(col);
      }
    }
  } else { // Multi row/column matrix
    for(let row = 0; row < matrix.length; row++) {
      for(let col = 0; col < matrix[0].length; col++) {
        if(matrix[row][col] === 1) {
          islandCount++;
          checkEdges(row, col);
        }
      }
    }
  }

  return islandCount;
}

// let testMatrix = [
//   [0, 1, 0, 1, 0],
//   [0, 0, 1, 1, 1],
//   [1, 0, 0, 1, 0],
//   [0, 1, 1, 0, 0],
//   [1, 0, 1, 0, 1]
// ];

let testMatrix;

testMatrix = [
  [1, 1, 1, 1, 0],
  [1, 1, 0, 1, 0],
  [1, 1, 0, 0, 0],
  [0, 0, 0, 0, 0]
]; // => 1 island

// testMatrix = [
//   [1, 1, 0, 0, 0],
//   [1, 1, 0, 0, 0],
//   [0, 0, 1, 0, 0],
//   [0, 0, 0, 1, 1]
// ]; // => 3 island

/*
LeetCode input tests
 */
//testMatrix = ["0", "0", "0"]; // => 1 island
//testMatrix = ["1011011"]; // => 3 island

//testMatrix = ["1","0", "1"];

let res = islandCount(testMatrix);
console.log("Result:", res);

File: interviewBit/island_count.py
# Island Count

"""
0  1  0  1  0
0  0  1  1  1
1  0  0  1  0
0  1  1  0  0
1  0  1  0  1
"""

def island_count(matrix):
    height = len(matrix)
    width = len(matrix[0])
    num_islands = 0

    def print_row(matrix):
        print_str = ""
        for row in range(height):
            row_str = ""
            for col in range(width):
                row_str+=str(matrix[row][col])
            print_str += row_str + "\n"
        print(print_str)

    def is_valid_position(matrix, row, col):
        if (0 < row < height) and (0 < col < width) and matrix[row][col] == 1: return True
        else: return False

    def expand_island(matrix, row, col):
        matrix[row][col] = 2
        # Check Up
        if (is_valid_position(matrix, row - 1, col)): expand_island(matrix, row-1, col)
        # Check Down
        if (is_valid_position(matrix, row + 1, col)): expand_island(matrix, row + 1, col)
        # Check Left
        if (is_valid_position(matrix, row, col - 1)): expand_island(matrix, row, col - 1)
        # Check Right
        if (is_valid_position(matrix, row, col + 1)): expand_island(matrix, row, col + 1)

    print("Before expanding:")
    print_row(matrix)

    # Scan row until 1 found, then expand island
    for row in range(height):
        for col in range(width):
            if(matrix[row][col] == 1):
                num_islands+=1
                expand_island(matrix, row, col) # Expand
                print("After expanding:")
                print_row(matrix)

    return num_islands



"""
Testing
"""
test_island = [
    [0, 1, 0, 1, 0],
    [0, 0, 1, 1, 1],
    [1, 0, 0, 1, 0],
    [0, 1, 1, 0, 0],
    [1, 0, 1, 0, 1],
]

result = island_count(test_island)
print ("NumIslands result:%d" % result)




